crypto/cmp/cmp_msg.c

```diff
@@ -0,0 +1,993 @@
+            goto err;
+        return msg;
+    case OSSL_CMP_PKIBODY_PKICONF:
+        if ((msg->body->value.pkiconf = ASN1_TYPE_new()) == NULL)
+            goto err;
+        ASN1_TYPE_set(msg->body->value.pkiconf, V_ASN1_NULL, NULL);
+        return msg;
+
+    case OSSL_CMP_PKIBODY_POLLREQ:
+        if ((msg->body->value.pollReq = sk_OSSL_CMP_POLLREQ_new_null()) == NULL)
+            goto err;
+        return msg;
+    case OSSL_CMP_PKIBODY_POLLREP:
+        if ((msg->body->value.pollRep = sk_OSSL_CMP_POLLREP_new_null()) == NULL)
+            goto err;
+        return msg;
+
+    case OSSL_CMP_PKIBODY_GENM:
+    case OSSL_CMP_PKIBODY_GENP:
+        if ((msg->body->value.genm = sk_OSSL_CMP_ITAV_new_null()) == NULL)
+            goto err;
+        return msg;
+
+    case OSSL_CMP_PKIBODY_ERROR:
+        if ((msg->body->value.error = OSSL_CMP_ERRORMSGCONTENT_new()) == NULL)
+            goto err;
+        return msg;
+
+    default:
+        CMPerr(0, CMP_R_UNEXPECTED_PKIBODY);
+        goto err;
+    }
+
+ err:
+    OSSL_CMP_MSG_free(msg);
+    return NULL;
+}
+
+#define HAS_SAN(ctx) \
+    (sk_GENERAL_NAME_num((ctx)->subjectAltNames) > 0 \
+         || OSSL_CMP_CTX_reqExtensions_have_SAN(ctx) == 1)
+
+static X509_NAME *determine_subj(OSSL_CMP_CTX *ctx, X509 *refcert,
+                                 int bodytype)
+{
+    if (ctx->subjectName != NULL)
+        return ctx->subjectName;
+
+    if (refcert != NULL
+            && (bodytype == OSSL_CMP_PKIBODY_KUR || !HAS_SAN(ctx)))
+        /*
+         * For KUR, copy subjectName from reference certificate.
+         * For IR or CR, do the same only if there is no subjectAltName.
+         */
+        return X509_get_subject_name(refcert);
+    return NULL;
+}
+
+/*
+ * Create CRMF certificate request message for IR/CR/KUR
+ * returns a pointer to the OSSL_CRMF_MSG on success, NULL on error
+ */
+static OSSL_CRMF_MSG *crm_new(OSSL_CMP_CTX *ctx, int bodytype,
+                              int rid, EVP_PKEY *rkey)
+{
+    OSSL_CRMF_MSG *crm = NULL;
+    X509 *refcert = ctx->oldCert != NULL ? ctx->oldCert : ctx->clCert;
+    /* refcert defaults to current client cert */
+    STACK_OF(GENERAL_NAME) *default_sans = NULL;
+    X509_NAME *subject = determine_subj(ctx, refcert, bodytype);
+    int crit = ctx->setSubjectAltNameCritical || subject == NULL;
+    /* RFC5280: subjectAltName MUST be critical if subject is null */
+    X509_EXTENSIONS *exts = NULL;
+
+    if (rkey == NULL
+            || (bodytype == OSSL_CMP_PKIBODY_KUR && refcert == NULL)) {
+        CMPerr(0, CMP_R_INVALID_ARGS);
+        return NULL;
+    }
+    if ((crm = OSSL_CRMF_MSG_new()) == NULL)
+        return NULL;
+    if (!OSSL_CRMF_MSG_set_certReqId(crm, rid)
+        /*
+         * fill certTemplate, corresponding to CertificationRequestInfo
+         * of PKCS#10. The rkey param cannot be NULL so far -
+         * it could be NULL if centralized key creation was supported
+         */
```

Comment:
indentation wrong.

Reply:
We do not see any indentation issue in this code section nor around it.
Also my new format checking tool #10363 does not complain here.

Reply:
I think this comment applies to the following term which is indented by 4-spaces,
so this comment should be indented as well.

Reply:
Ah, right!
I did not notice this because I did not read the comment contents, and of course the tool cannot find this either because it cannot understand whether a multi-line comment pertains to the line(s) before or after.
We'll fix this tomorrow.

Reply:
We were already wondering if this your only comment this time ;-)

Reply:
Thanks, I fixed this.